numpy/core/src/multiarray/nditer_constr.c

```diff
@@ -1410,6 +1413,59 @@ check_mask_for_writemasked_reduction(NpyIter *iter, int iop)
     return 1;
 }
 
+
+/*
+ * Check whether a reduction is OK based on the flags and the operand being
+ * readwrite. This path is deprecated, since usually only specific axis
+ * should be reduced. If axes are specified explicitely, the flag is
+ * not needed
+ */
+static int
+npyiter_check_reduce_ok_and_set_flags(
+        NpyIter *iter, npy_uint32 flags, npyiter_opitflags *op_itflags) {
+    /* If it's writeable, this means a reduction */
+    if (*op_itflags & NPY_OP_ITFLAG_WRITE) {
+        if (!(flags & NPY_ITER_REDUCE_OK)) {
+            PyErr_SetString(PyExc_ValueError,
+                    "output operand requires a reduction, but reduction is"
+                    "not enabled");
+            return 0;
+        }
+        if (!(*op_itflags & NPY_OP_ITFLAG_READ)) {
+            PyErr_SetString(PyExc_ValueError,
+                    "output operand requires a reduction, but is flagged as "
+                    "write-only, not read-write");
+            return 0;
+        }
+        NPY_IT_DBG_PRINT("Iterator: Indicating that a reduction is"
+                         "occurring\n");
+
+        NIT_ITFLAGS(iter) |= NPY_ITFLAG_REDUCE;
+        *op_itflags |= NPY_OP_ITFLAG_REDUCE;
+    }
+    return 1;
+}
```

Comment:
Would recommend returning `-1` for errors, not 0.
```suggestion
static int
npyiter_check_reduce_ok_and_set_flags(
        NpyIter *iter, npy_uint32 flags, npyiter_opitflags *op_itflags) {
    /* If it's writeable, this means a reduction */
    if (*op_itflags & NPY_OP_ITFLAG_WRITE) {
        if (!(flags & NPY_ITER_REDUCE_OK)) {
            PyErr_SetString(PyExc_ValueError,
                    "output operand requires a reduction, but reduction is"
                    "not enabled");
            return -1;
        }
        if (!(*op_itflags & NPY_OP_ITFLAG_READ)) {
            PyErr_SetString(PyExc_ValueError,
                    "output operand requires a reduction, but is flagged as "
                    "write-only, not read-write");
            return -1;
        }
        NPY_IT_DBG_PRINT("Iterator: Indicating that a reduction is"
                         "occurring\n");

        NIT_ITFLAGS(iter) |= NPY_ITFLAG_REDUCE;
        *op_itflags |= NPY_OP_ITFLAG_REDUCE;
    }
    return 0;
}
```

Reply:
Hmm, I see the rest of this file does this. Perhaps a cleanup patch to switch the meaning would be sensible, but perhaps it would just be churn.